### ISSUE TITLE
Firebird int comparison is cast to BigInt although dialiect 1

### ISSUE THREAD

**Issue by Franki1986** [linq2db/linq2db#1318]
LinqToDB-Version : 2.3.0
Firebird-Version: 2.1.3
SQL-Dialect: 1

Hello,

I run a simple SQL that searches for a specific number:

UInt32 number = 50;
using (var context = new DbContext())
               {
                  var p = context.Persons.FirstOrDefault(p1 => p1.Age == number)
               }

In  the resulting sql the Age property is cast to a BigInt. 
The mapped property is UInt16.

The problem is that BigInts are not supported by SQL-Dialect 1.

**Comment by Franki1986** [comment]
The [FirebirdSqlBuilder](https://github.com/linq2db/linq2db/blob/b182372bc62367a134670212d0ae92df3c2f9b30/Source/LinqToDB/DataProvider/Firebird/FirebirdSqlBuilder.cs#L89)  should convert UInt32 to integer if sql dialect is 1, otherwise the default [BuildDataType](https://github.com/linq2db/linq2db/blob/012c86c8c9b682126ad1aae4c431209e037c68ab/Source/SqlProvider/BasicSqlBuilder.cs#L2233) will convert the UInt32 to BigInt.

**Comment by Franki1986** [comment]
Is there a way so that the mapped property is not cast also  to a bigint?I mean the passed Parameter is covered with the above convert method, but the compared mapped property (does not matter if int or short)  is still cast to the original Parameter type (bigint)

**Comment by Franki1986** [comment]
I have made a pull request: [
58cd187](https://github.com/linq2db/linq2db/commit/58cd1879cfa6a34944a65f5c6b92e7b3dfe73564)

**Comment by MaceWindu** [comment]
Aww, another option... We really need to implement automatic dialect/version detection asap..

**Comment by Franki1986** [comment]
Yes, indeed. that would be the better way. To my question: Is there a workaround to overcome the automatic cast of the mapped property?

**Comment by Franki1986** [comment]
Any news how this will be fixed or for a workaround?

**Comment by MaceWindu** [comment]
I plan to look at it today/tomorrow so it will be included into next release

**Comment by MaceWindu** [comment]
@Franki1986 could you test dialect detection from this PR? https://github.com/linq2db/linq2db/pull/1334

**Comment by MaceWindu** [comment]
Moving it to 3.0 milestone as it looks like requested functionality could be implemented by configuring provider/mapping schema and proper dialects support is something we plan to implement in 3+ version.
